ckanext/xloader/jobs.py

```diff
@@ -159,57 +162,61 @@ def xloader_data_into_datastore_(input, job_dict):
-                )
-        response = get_url()
-
-        if response.status_code == 202:
-            # Seen: https://data-cdfw.opendata.arcgis.com/datasets
-            # In this case it means it's still processing, so do retries.
-            # 202 can mean other things, but there's no harm in retries.
-            wait = 1
-            while wait < 120 and response.status_code == 202:
-                logger.info('Retrying after {}s'.format(wait))
-                time.sleep(wait)
-                response = get_url()
-                wait *= 3
-        response.raise_for_status()
+        response = get_response(url, headers)
 
         cl = response.headers.get('content-length')
         if cl and int(cl) > MAX_CONTENT_LENGTH:
-            error_msg = 'Resource too large to download: ' \
-                '{cl} > max ({max_cl}).' \
-                .format(cl=cl, max_cl=MAX_CONTENT_LENGTH)
-            logger.warning(error_msg)
-            raise JobError(error_msg)
+            raise DataTooBigError()
 
         # download the file to a tempfile on disk
-        filename = url.split('/')[-1].split('#')[0].split('?')[0]
-        tmp_file = tempfile.NamedTemporaryFile(suffix=filename)
-        length = 0
-        m = hashlib.md5()
         for chunk in response.iter_content(CHUNK_SIZE):
             length += len(chunk)
             if length > MAX_CONTENT_LENGTH:
-                raise JobError(
-                    'Resource too large to process: {cl} > max ({max_cl}).'
-                    .format(cl=length, max_cl=MAX_CONTENT_LENGTH))
+                raise DataTooBigError
             tmp_file.write(chunk)
             m.update(chunk)
+        data['datastore_contains_all_records_of_source_file'] = True
 
+    except DataTooBigError:
+        tmp_file.close()
+        cl = response.headers.get('content-length')
+        message = 'Data too large to load into Datastore: ' \
+                    '{cl} bytes > max {max_cl} bytes.' \
+                    .format(cl=cl or length, max_cl=MAX_CONTENT_LENGTH)
+        logger.warning(message)
+        file_format = resource.get('format')
+        if MAX_EXCERPT_LINES > 0:
+            if file_format not in EXCERPT_FORMATS:
+                logger.info('Loading excerpt for {format} not supported.'.format(format=file_format))
+                raise JobError(message)
+            logger.info('Loading excerpt of ~{max_lines} lines to '
+                        'DataStore.'
+                        .format(max_lines=MAX_EXCERPT_LINES))
+            tmp_file = get_tmp_file(url)
+            response = get_response(url, headers)
+            length = 0
+            line_count = 0
+            m = hashlib.md5()
+            for line in response.iter_lines(CHUNK_SIZE):
+                tmp_file.write(line + '\n')
+                m.update(line)
+                length += len(line)
+                line_count += 1
+                if length > MAX_CONTENT_LENGTH or line_count == MAX_EXCERPT_LINES:
```

Comment:
Not really an issue, but I would use `line_count >= MAX_EXCERPT_LINES` just to be safe. Right now it's not possible to get past this, but if for whatever reason you get two lines at once, this check would fail.